### PR TITLE
Update NES and SNES hashing methods

### PR DESCRIPTION
**Pull request by rzumer**
## Description

VROM was discarded from the content hashed in the NES strategy for historical reasons, but full content hashes are now available in the database, so this method is deprecated and full content hashing (excluding the header) is now preferred.

This is tested and matches https://github.com/RetroAchievements/RAEmus/commit/67884716c8e8d14d4a744ff49717982c4cd47bfd.

In addition, SNES headers should be skipped on hashing. This attempts to detect a headered file based on file size, and falls back to `GENERIC_MD5` if it does not match an expected pattern, as is done with NES.

## Reviewers

@leiradel

### CODE REVIEW

**Comment by retro-wertz** [comment]
this is incomplete. PR still does not add vrom data for hashing. this is only removing the computation for prg rom size for mappers that are not  powerof2's (mappers 53 198 228 known so far to only not be power of 2)

**Comment by rzumer** [comment]
@retro-wertz Thank you for checking, it should work now (tested again on my end).

**Comment by twinaphex** [comment]
@retro-wertz @leiradel Should I merge this?

**Comment by leiradel** [comment]
It looks ok, I'll take a look later because I think there's an optimization that we can make.

**Comment by leiradel** [comment]
It's good, it simply fails to get the game ID if the content doesn't have a NES header. In this case, the `GENERIC_MD5` heuristic will evaluate the MD5 of the entire content and get the game ID from the server.

**Comment by retro-wertz** [comment]
PR should behave the same previously, other than adding crc for the chr part if rom has it, so its good. atm only mesen is able to support non-headered roms

**Comment by meleu** [comment]
hello guys,
about a week ago I was working on a PR for this same subject with the intention to make RetroArch have the same behavior as the official standalone RAEmus, and then faced some RL issues...

desired behavior for **NES**:
If the ROM starts with "NES\1a", the first 16 bytes are ignored and the remaining file contents are hashed. If the ROM does not start with "NES\1a", the entire file is hashed.

my commit: c9035524e8f893d50882ee31ffe5c52c50eba29e


desired behavior for **SNES**:
If the size of the file is 512 bytes more than a multiple of 8K, the first 512 bytes are ignored and the remaining file contents are hashed. If the size of the file is not 512 bytes more than a multiple of 8K, the entire file is hashed.

my commit: ba0b0f4cf1f58bc69dc216fa2c50781537dd300d


I've been busy lately but if you guys agree I can submit a PR using my branch with those commits.

https://github.com/meleu/RetroArch/tree/patch-1

**Comment by rzumer** [comment]
@meleu For NES, your version processes non-NES files twice (full content hashing and server request). This PR skips the NES method when no header is present and falls back to full content hashing, which should be more efficient.

I don't know much about SNES, so with that in mind I suggest you send another PR with just those changes.

File: cheevos/cheevos.c [hunk 1]
```diff
@@ -2902,24 +2902,21 @@ static int cheevos_iterate(coro_t *coro)
...
+          coro->gameid = 0;
+          CORO_RET();
+      }
 
+      coro->offset = 512;
       coro->count  = 0;
 
       CORO_GOSUB(EVAL_MD5);
-
-      if (coro->count == 0)
```
Comment (rzumer) [inline cheevos/cheevos.c]: I removed this part because it seemed redundant. I would like to add it back with a comment if it is useful.

**Comment by rzumer** [comment]
Updated NES method to match https://github.com/RetroAchievements/RALibretro/pull/50, which should also match the method described by @meleu earlier.

**Comment by retro-wertz** [comment]
how does it know to ignore the trainer data, if header is not check for this 512-byte data before the PRG section starts?

**Comment by rzumer** [comment]
> 
> 
> how does it know to ignore the trainer data, if header is not check for this 512-byte data before the PRG section starts?

Poor choice of words maybe, but it ignores the fact that trainer data exists at all.

**Comment by retro-wertz** [comment]
?

**Comment by rzumer** [comment]
> 
> 
> ?

Here "ignore trainer data" means "do not care about trainer, hash everything except NES header". Is that clearer?

**Comment by rzumer** [comment]
> 
> ignoring (and in this case totally removing coro->header.rom_type & 4's check) will add these 512 bytes of data into the hash which is not suppose to happen
> 

From what I understand, it *is* supposed to happen, in order to block trainers from accessing RetroAchievements.

**Comment by retro-wertz** [comment]
i dont think they add the trainer data... by your original PR linked, they are only passing PRG and CHR data to RA_OnLoadNewRom()

https://github.com/RetroAchievements/RAEmus/commit/67884716c8e8d14d4a744ff49717982c4cd47bfd

**Comment by rzumer** [comment]
@retro-wertz Please refer to the other link I posted earlier.
I wish that the reference implementations were both changed to match, but clearly someone forgot to do so for that one.

**Comment by Jamiras** [comment]
rzumer is correct. Feedback was provided when attempting to make the change in RALibRetro that was not applied to RANes. I will make the appropriate changes there soon.

For reference, I've documented the various hashing methods for each system (as I understand them) here: http://docs.retroachievements.org/Game-Identification/

File: cheevos/cheevos.c [hunk 2]
```diff
@@ -2902,30 +2902,23 @@ static int cheevos_iterate(coro_t *coro)
...
-
-      CORO_GOSUB(EVAL_MD5);
-
-      if (coro->count == 0)
+      /* Checks for the existence of a headered SNES file.
+         Unheadered files fall back to GENERIC_MD5. */
+      
+      const int SNES_HEADER_LEN = 0x200;
+      
+      if (coro->len % 0x2000 != SNES_HEADER_LEN || coro->len <= SNES_HEADER_LEN)
```
Comment (Jamiras) [inline cheevos/cheevos.c]: The second clause is probably unnecessary. If len < SNES_HEADER_LEN, it won't modulus to SNES_HEADER_LEN, and if it is exactly SNES_HEADER_LEN, the file won't match using either the truncated for full file hash.

**Comment by retro-wertz** [comment]
again its incorrect to hash the PRG data together with the trainer data, and the CHR together with tiled data. i dont know why they want to do it like that. if the trainer data and the extra tiled data at the end of chr is needed, i could understand passing them as separate variables, but since this how Retroachievement is doing it, then i have no more complains...

**Comment by Jamiras** [comment]
For our purposes, we're trying to ensure the player is using exactly the ROM the set was designed for. Having a trainer attached may alter the behavior of the ROM and cause issues with the achievement triggers. Stripping the header should theoretically increase support as both headered and non-headered ROMs are now viable. The original change to include the CHR data also ensures that our hashes match hashes for ROMs available on the internet, which should reduce the number of questions we get about finding the "right" ROM. If the player has a ROM with a trainer, their hash _won't_ match, and they'll go look for one without the trainer.

File: cheevos/cheevos.c (hunk 2 above)
Comment (rzumer) [reply cheevos/cheevos.c]: It is better to avoid sending unnecessary requests.

Comment (meleu) [reply cheevos/cheevos.c]: just a comment about this:

> All hashing methods are actually called for every file, regardless of core

I may be wrong, but I think the heuristic to figure the system being emulated on the cheevos context is based on the content's file extension.

And it's done only for some known SNES, Genesis and Lynx extensions. For example, SNES method will only be used if the content's file extension is one of these: 
`smc`, `fig`, `sfc`, `gd3`, `gd7`, `dx2`, `bsx`, `swc`, as we can see [in this part of the code](https://github.com/libretro/RetroArch/blob/ab2010c4940db91278ae2f0ca88c13e55e3f2065/cheevos/cheevos.c#L2594-L2605).

This snippet makes me believe that later methods will be called only if the previous ones fail:
```c
   static cheevos_finder_t finders[] =
   {
      {SNES_MD5,    "SNES (8Mb padding)",                snes_exts},
      {GENESIS_MD5, "Genesis (6Mb padding)",             genesis_exts},
      {LYNX_MD5,    "Atari Lynx (only first 512 bytes)", lynx_exts},
      {NES_MD5,     "NES (discards VROM)",               NULL},
      {GENERIC_MD5, "Generic (plain content)",           NULL},
      {FILENAME_MD5, "Generic (filename)",               NULL}
};
```

Therefore the worst case seems to be
(1) `SNES_MD5` **or** `GENESIS_MD5` **or** `LYNX_MD5` (based on the content's file extension), and then (2) `NES_MD5`, and then (3) `GENERIC_MD5`, and then (4) `FILENAME_MD5`. Which means 4 tries before total failure.

Comment (rzumer) [reply cheevos/cheevos.c]: You are right, but it is possible for someone to load a file with an extension in `snes_exts`, so why not make sure that it is worth a request before doing it? It is a cheap test.

Comment (rzumer) [reply cheevos/cheevos.c]: Updated as per my comment earlier in this thread.

**Comment by meleu** [comment]
hey @twinaphex , from the RA staff point of view this PR is ready to be merged.

**Comment by leiradel** [comment]
This looks good judging from the diff.

**Pull request merged**